sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/MapElements.java

```diff
@@ -85,42 +85,57 @@ private MissingOutputTypeDescriptor(SerializableFunction<InputT, OutputT> fn) {
       this.fn = fn;
     }
 
-    public MapElements<InputT, OutputT> withOutputType(TypeDescriptor<OutputT> outputType) {
-      return new MapElements<>(fn, outputType);
+    public MapElements<InputT, OutputT> withOutputType(final TypeDescriptor<OutputT> outputType) {
+      return new MapElements<>(
+          SimpleFunction.fromSerializableFunctionWithOutputType(fn, outputType), fn.getClass());
     }
+
   }
 
   ///////////////////////////////////////////////////////////////////
 
-  private final SerializableFunction<InputT, OutputT> fn;
-  private final transient TypeDescriptor<OutputT> outputType;
+  private final SimpleFunction<InputT, OutputT> fn;
+  private final Class<?> fnClass; // for display data purposes
```

Comment:
Careful, storing `Class<?>` instances can be problematic. Anonymous classes backed by Java 8 lambdas will explode during serialization. See [CombineJava8Test#testLambdaSerialization()](https://github.com/apache/incubator-beam/blob/master/sdks/java/java8tests/src/test/java/org/apache/beam/sdk/transforms/CombineJava8Test.java#L144). In Combine.java, we [store the `DisplayData.Item<?>` instead](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java#L127).

Recommend writing a Java8 test for this class as well. I wonder: would be it be easy to write some FindBugs-like static analysis that would catch `Serializable` classes with non-transient `Class<?>` instance fields?